src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -666,10 +698,97 @@ MetadataCreateCommands(void)
 												  shardCreateCommandList);
 	}
 
+	/* As the last step, propagate the pg_dist_object entities */
+	List *distributedObjectSyncCommandList = DistributedObjectSyncCommandList();
+	metadataSnapshotCommandList = list_concat(metadataSnapshotCommandList,
+											  distributedObjectSyncCommandList);
+
 	return metadataSnapshotCommandList;
 }
 
 
+/*
+ * DistributedObjectSyncCommandList returns the necessary commands to create
+ * pg_dist_object entries on the new node.
+ */
+static List *
+DistributedObjectSyncCommandList(void)
+{
+	HeapTuple pgDistObjectTup = NULL;
+	Relation pgDistObjectRel = table_open(DistObjectRelationId(), AccessShareLock);
+	Relation pgDistObjectIndexRel = index_open(DistObjectPrimaryKeyIndexId(),
+											   AccessShareLock);
+	TupleDesc pgDistObjectDesc = RelationGetDescr(pgDistObjectRel);
+
+	List *objectAddresses = NIL;
+	List *distributionArgumentIndexes = NIL;
+	List *colocationIds = NIL;
+
+	SysScanDesc pgDistObjectScan = systable_beginscan_ordered(pgDistObjectRel,
+															  pgDistObjectIndexRel, NULL,
+															  0, NULL);
+	while (HeapTupleIsValid(pgDistObjectTup = systable_getnext_ordered(pgDistObjectScan,
+																	   ForwardScanDirection)))
+	{
+		Form_pg_dist_object pg_dist_object = (Form_pg_dist_object) GETSTRUCT(
+			pgDistObjectTup);
+
+		ObjectAddress *address = palloc(sizeof(ObjectAddress));
+		int32 *colocationIdPtr = palloc(sizeof(int32));
+		int32 *distributionArgumentIndexPtr = palloc(sizeof(int32));
+
+		ObjectAddressSubSet(*address, pg_dist_object->classid, pg_dist_object->objid,
+							pg_dist_object->objsubid);
+
+		bool distributionArgumentIndexIsNull = false;
+		Datum distributionArgumentIndexDatum =
+			heap_getattr(pgDistObjectTup,
+						 Anum_pg_dist_object_distribution_argument_index,
+						 pgDistObjectDesc,
+						 &distributionArgumentIndexIsNull);
+		*distributionArgumentIndexPtr = DatumGetInt32(distributionArgumentIndexDatum);
+
+		bool colocationIdIsNull = false;
+		Datum colocationIdDatum =
+			heap_getattr(pgDistObjectTup,
+						 Anum_pg_dist_object_distribution_argument_index,
+						 pgDistObjectDesc,
+						 &colocationIdIsNull);
+		*colocationIdPtr = DatumGetInt32(colocationIdDatum);
+
+		objectAddresses = lappend(objectAddresses, address);
+
+		if (distributionArgumentIndexIsNull)
+		{
+			distributionArgumentIndexes = lappend(distributionArgumentIndexes, NULL);
+		}
+		else
+		{
+			distributionArgumentIndexes = lappend(distributionArgumentIndexes,
+												  distributionArgumentIndexPtr);
+		}
+
+		if (colocationIdIsNull)
+		{
+			colocationIds = lappend(colocationIds, NULL);
+		}
+		else
+		{
+			colocationIds = lappend(colocationIds, colocationIdPtr);
+		}
+	}
+
+	char *workerMetadataUpdateCommand = MarkObjectsDistributedCreateCommand(
+		objectAddresses, distributionArgumentIndexes, colocationIds);
+	List *commandList = list_make1(workerMetadataUpdateCommand);
+
+	systable_endscan_ordered(pgDistObjectScan);
```

Comment:
minor: we should be able to move `MarkObjectsDistributedCreateCommand` after the scan is over?